storage/isolation_manager.rs

```diff
@@ -241,297 +358,317 @@ struct Timeline {
+        assert!(end_relative_index >= relative_index_of_window_0);
+        let end_lies_in_window = (end_relative_index - relative_index_of_window_0) as usize / TIMELINE_WINDOW_SIZE;
+        let mut concurrent_windows: Vec<Arc<TimelineWindow<TIMELINE_WINDOW_SIZE>>> = Vec::new();
+
+        let window_after = (0..windows.len()).find(|i| {
+            let window_start_seq = windows.get(*i).unwrap().starting_sequence_number().unwrap().number();
+            window_start_seq > start_sequence_number.number()
+        });
+        let first_concurrent_window = match window_after {
+            None => windows.len() - 1,
+            Some(i) => {
+                if i == 0 {
+                    i
+                } else {
+                    i - 1
+                }
+            }
+        };
+        for i in first_concurrent_window..(end_lies_in_window + 1) {
+            concurrent_windows.push(windows.get(i).unwrap().clone());
+        }
+        let start_index_of_first_concurrent =
+            relative_index_of_window_0 + (first_concurrent_window * TIMELINE_WINDOW_SIZE) as i64;
+        (concurrent_windows, start_index_of_first_concurrent)
     }
 
-    fn get_window(&self, sequence_number: SequenceNumber) -> Arc<TimelineWindow<TIMELINE_WINDOW_SIZE>> {
-        self.try_get_window(sequence_number).unwrap()
-    }
+    fn try_get_window(&self, relative_index: i64) -> Option<(Arc<TimelineWindow<TIMELINE_WINDOW_SIZE>>, usize)> {
+        let (next_window, windows) = &*self.next_window_and_windows.read().unwrap_or_log();
+        let start_index: i64 = next_window - (windows.len() * TIMELINE_WINDOW_SIZE) as i64;
+        let must_be_in_window = (relative_index - start_index) as usize / TIMELINE_WINDOW_SIZE;
 
-    fn try_get_window(&self, sequence_number: SequenceNumber) -> Option<Arc<TimelineWindow<TIMELINE_WINDOW_SIZE>>> {
-        let (_, windows) = &*self.next_window_and_windows.read().unwrap_or_log();
-        Self::find_window(sequence_number, windows)
+        if must_be_in_window >= 0 && must_be_in_window < windows.len() {
+            let in_window = windows.get(must_be_in_window)?;
+            let slot_index: usize = (relative_index - start_index) as usize % TIMELINE_WINDOW_SIZE;
+            Some((in_window.clone(), slot_index))
+        } else {
+            None
+        }
     }
 
-    fn find_window(
-        sequence_number: SequenceNumber,
-        windows: &VecDeque<Arc<TimelineWindow<TIMELINE_WINDOW_SIZE>>>,
-    ) -> Option<Arc<TimelineWindow<TIMELINE_WINDOW_SIZE>>> {
-        windows.iter().rev().find(|window| window.contains(sequence_number)).cloned()
+    fn get_or_create_window(&self, relative_index: i64) -> (Arc<TimelineWindow<TIMELINE_WINDOW_SIZE>>, usize) {
+        let create: bool = {
+            let (next_window, _) = &*self.next_window_and_windows.read().unwrap_or_log();
+            relative_index >= *next_window
```

Comment:
this boolean expression can be moved into the `if` so we don't have to look so far away for the logic